### PR TITLE
Replace intltools-merge with msgfmt

### PR DESCRIPTION
**Pull request by Nick-Hall**
The build now requires gettext v0.19.7 or above.
    
Removed obsolete files:  gramps.applications,  gramps.keys and gramps.mime
    
Created ITS rules for holidays.xml and tips.xml files.
    
Included ITS files for shared-mime-info from a gettext v0.20.2 installation.

### CODE REVIEW

File: data/gramps.desktop.in [hunk 1]
```diff
@@ -1,13 +1,13 @@
 [Desktop Entry]
-_Name=Gramps
-_GenericName=Genealogy System
-_X-GNOME-FullName=Gramps Genealogy System
-_Comment=Manage genealogical information, perform genealogical research and analysis
+Name=Gramps
```
[the remark this answers is not in the document]
Comment (Nick-Hall) [reply data/gramps.desktop.in:2]: I've rewritten these to use [ITS](https://www.w3.org/TR/its20/) rules.  There are some examples [here](https://www.gnu.org/software/gettext/manual/html_node/Preparing-ITS-Rules.html) in the gettext manual.

**Comment by Nick-Hall** [comment]
The Travis CI build is now passing.

**Comment by DavidMStraub** [comment]
Hi all,

sorry in case this is a bit off-topic, but I guess it's related to this discussion:

I was working on trying to build Gramps wheels for different platforms to be able to provide them on PyPI at some point (my main motivation being the [web API](https://github.com/gramps-project/web-api), but also related to this [feature request](https://gramps-project.org/bugs/view.php?id=11404) and this [pull request](https://github.com/gramps-project/gramps/pull/1106)). Here, `gettext` is a major annoyance, especially when building for Windows or Mac OS. So I was wondering why we don't use `msgfmt.py` [provided with Python](https://github.com/python/cpython/blob/master/Tools/i18n/msgfmt.py) itself. I then found a [comment](https://github.com/gramps-project/gramps/blob/master/windows/nonAIO/builder/build_GrampsWin32.py#L226-L227) in a Windows build script "no longer using python msgfmt as it doesn't handle plurals (april 2010)". However apparently this support has been added later in 2010 [(see commit)](https://github.com/python/cpython/commit/5aafc17405127b728c66b2e7e516c1c9c4d660cb), even if it seems they forgot to update the header comment.

To cut a long story short: would it by possible to use `msgfmt.py` in all the builds instead of gettext? It would strongly simplify building wheels for different platforms.

(Disclaimer: just to be clear, these wheels are not meant to replace the existing packaging formats but would rather be used when Gramps servers as a library, e.g. for web apps or Jupyter notebooks.)

**Comment by Nick-Hall** [comment]
> To cut a long story short: would it by possible to use msgfmt.py in all the builds instead of gettext?

No, because the python version doesn't support XML or Desktop files.

It may be possible to use it in a core distribution that doesn't need the _desktop_, _appdata_ or _shared-mime-info_ files.

**Comment by jralls** [comment]
> No, because the python version doesn't support XML or Desktop files.

@Nick-Hall I think you're confusing `msgfmt` and `xgettext`. The former is interested *only* in po files: It checks them, counts the translations in them, and most important in this context compiles them into mo files. It cares not a whit for what kind of file the `msgstr`s come from. As long as `msgfmt.py` can do that last task it would be fine for the `build_trans` target in `setup.py`.

`xgettext` is for extracting the translatable strings and its pure-python stand-in `pygettext` is indeed not useful to us because it can handle only python source files. It's also not routinely required: Although it can certainly be run more often it's only really necessary along with an immediately following run of `msgmerge` before generating release tarballs.

**Comment by Nick-Hall** [comment]
> As long as `msgfmt.py` can do that last task it would be fine for the `build_trans` target in `setup.py`.

I agree, but this PR now uses `msgfmt` for the `trans_intl` target with the `--xml` and `--desktop` options to replace `intltool_merge`.

**Comment by jralls** [comment]
> I agree, but this PR now uses msgfmt for the trans_intl target with the --xml and --desktop options to replace intltool_merge.

Maybe that could be optional? It's only useful on Linux and BSD desktops, nothing else uses those files.

**Comment by Nick-Hall** [comment]
Yes.  I was considering a `--core` or `--cli` option.

I notice that `msgfmt.py` is installed as part of my python package (v3.8.5).  When I run it with `--help` it states:  "Currently it
does not handle plural forms but it does handle message contexts.".

**Comment by jralls** [comment]
The without-(xml|desktop) isn't really "core", more like "not Linux Desktop". If you call it "core" someone will come along later and decide that it should exclude building the GUI.

If you search `msgfmt.py` for 'plural' you'll find that it was in fact implemented (`git blame` shows it 10 years ago) but nobody fixed the comment at the top of the file.

**Comment by romjerome** [comment]
> "Currently it does not handle plural forms but it does handle message contexts.".

As Gramps still uses "old" message context format (inherited from gnome), maybe
it can pass ? https://gramps-project.org/bugs/view.php?id=3416

**Comment by DavidMStraub** [comment]
> If you search msgfmt.py for 'plural' you'll find that it was in fact implemented (git blame shows it 10 years ago) but nobody fixed the comment at the top of the file.

Indeed, looks like a great opportunity to contribute a pull request to CPython :wink: 

>  I was considering a --core or --cli option.

Concerning the `pip` install from PyPI: given the difficulty of building platform dependent wheels for all necessary platforms, when actually the `.po` files are the only binary files in the distribution, it currently looks to me like it would be better to simply upload the source distribution to PyPI. However, an important point to note is that, in this case, `pip` will essentially call `python setup.py build` **without any command line options**. Thus my suggestion would be that calling `setup.py` without additional options will call the "minimal build" (only compile `.po` files, but don't build desktop/XML files).

**Comment by Nick-Hall** [comment]
> The without-(xml|desktop) isn't really "core", more like "not Linux Desktop". If you call it "core" someone will come along later and decide that it should exclude building the GUI.

Yes.  I was also considering removing the GUI code from the core build.  However, such a change would be beyond the scope of this PR.

> If you search msgfmt.py for 'plural' you'll find that it was in fact implemented (git blame shows it 10 years ago) but nobody fixed the comment at the top of the file.

I just compiled a couple of _po_ files using the gettext tools `msgfmt` and python equivalent.  The python version doesn't generate the hash table, but this isn't used in the python gettext module python anyway.  The only other difference was that the python po files were always 41 bytes larger - the gettext tools version excludes the POT create date in the header information.  For our purposes the two files are equivalent.

I'll probably just use the python version as a fallback and print a warning about the missing desktop files.

**Comment by Nick-Hall** [comment]
@romjerome Feature request #3416 looks interesting.  It may be possible to implement in earlier versions of python.

**Comment by DavidMStraub** [comment]
Two more comments after looking at this in more detail:

1. If you plan to use `msgfmt.py`, I suggest to use a copy of the most recent version and ship it with Gramps. This is because the install location of that script can differ from system to system and because older versions of Python might ship an outdated version (without plural support, for instance).
2. I realized that my initial reasoning:

> I was working on trying to build Gramps wheels for different platforms to be able to provide them on PyPI at some point [...]. Here, gettext is a major annoyance, especially when building for Windows or Mac OS. So I was wondering why we don't use msgfmt.py provided with Python itself.

was probably flawed: I was assuming that the `.mo` files are platform dependent. However, after looking at the source of `msgfmt.py`, it seems to me that `.mo` files compiled on some Linux should actually work anywhere. If this is the case (would be great if someone could confirm), it doesn't really matter if `gettext` is a build dependency since we can just compile the `.mo` files once and ship them as part of a *platform independent, pure-Python* wheel.

**Comment by jralls** [comment]
The context support is more of a potential issue (we don't use it ATM) given that plural support has been around for long enough that even very old Python versions will probably work. On the other hand we'd have to back-test the current release of msgfmt.py to make sure that it doesn't use any Python features that would cause it to fail with an older Python version.

I wonder, though, about how useful a pure-python Gramps wheel would be. Gramps has a bunch of C-library dependencies and associated PyPI packages that are partly C and partly Python so any end user is going to have to have MSYS2 on Windows or MacPorts or gtk-osx setup on a Mac to be able to build the dependencies. (I don't know if it's possible to get everything Gramps needs from Homebrew.)

**Comment by DavidMStraub** [comment]
> I wonder, though, about how useful a pure-python Gramps wheel would be. Gramps has a bunch of C-library dependencies and associated PyPI packages that are partly C and partly Python so any end user is going to have to have MSYS2 on Windows or MacPorts or gtk-osx setup on a Mac to be able to build the dependencies. (I don't know if it's possible to get everything Gramps needs from Homebrew.)

That's of course true, but as I stressed above:

 > Just to be clear, these wheels are not meant to replace the existing packaging formats but would rather be used when Gramps serves as a library, e.g. for web apps or Jupyter notebooks.

That's one of the awesome things about Gramps: to me it's not just a Desktop application, but also a very powerful Python library for exploring and managing a genealogical research database.

**Comment by jralls** [comment]
Yeah, I understand that. It's not just the GUI: Among those C dependencies are the database engines, SQLite3 and BerkeleyDB, needed to access Gramps's data. What good would a webapp be if it can't load the database?

**Comment by DavidMStraub** [comment]
`sqlite3` is part of the standard library, BSDDB will be deprecated in Gramps 5.2 AFAIK, so no problem there.

If you're curious, I also looked at the optional dependencies:
- `Pillow`: has wheels for all platforms, no problem
- `PyICU`: [no wheels unfortunately](https://github.com/ovalhub/pyicu/issues/119). On Ubuntu, requires `sudo apt install libicu-dev` (and a compiler). Haven't tried other platforms. 

The other dependencies (`PyGObject` or `pycairo`) are only relevant for the GUI.

**Comment by Nick-Hall** [comment]
If `msgfmt.py` is in the path, then using it instead of the gettext version of `msgfmt` appears to work.  However, I have decided to leave this for a separate PR.

**Pull request merged**